### PR TITLE
Add `canonize` function

### LINKED ISSUE #22: Implement Vidal-form canonization

### PR DESCRIPTION
**Pull request by jofrevalles**
# Summary
This PR resolves #22 by implementing the `canonize` function, which takes an input `Chain` and transforms it into the canonical form (also known as Vidal form), where we have a set of `Γ` and `Λ` tensors. For now, this function is implemented for `Chain` with `Open` boundaries.

Additionally, we added some tests to cover for the new function. Thanks to @starsfordummies for the useful comments!

# Example
Let's show how this function works:
```julia
julia> qtn = rand(Chain, Open, State; n=5, p = 2, χ=20)
MPS (inputs=0, outputs=5)

julia> canonized = canonize(qtn)
MPS (inputs=0, outputs=5)

julia> length(tensors(canonized)) # 5 tensors + 4 singular values vectors
9

julia> isapprox(contract(transform(TensorNetwork(canonized), Tenet.HyperindConverter())),
                contract(TensorNetwork(qtn)))
true

julia> Γ₃ = select(canonized, :tensor, Site(3))
4×2×4 Tensor{Float64, 3, Array{Float64, 3}}: ...

julia> Λ₄ = pop!(TensorNetwork(canonized), select(canonized, :between, Site(3), Site(4)))
4-element Tensor{Float64, 1, Vector{Float64}}:
 0.8303349576858878
 0.4869563863594485
 0.22360345538771664
 0.15303212266448157

julia> sum((i -> i^2).(Λ₄))
1.000000000000001

julia> replace!(TensorNetwork(canonized), Γ₃ => contract(Γ₃, Λ₄; dims=()))
TensorNetwork (#tensors=8, #inds=9)

julia> @test Qrochet.isrightcanonical(canonized, Site(3))
Test Passed
```

### CODE REVIEW

**Comment by jofrevalles** [comment]
Question for @bsc-quantic/tensor-networks: should we do that the `isleftcanonical`, `isrightcanonical` functions already process the singular values?

I think that makes sense that we do not (so like it is currently in the code), and just take the tensor at that `site` and do the checking. This makes the tests for this function a little bit cumbersome, but do the work. In my opinion, we should add a function that does the contraction of the singular values, since I think this will be necessary for many algorithms.

File: test/Ansatz/Chain_test.jl [hunk 1]
```diff
@@ -147,6 +147,45 @@
...
 
+        @testset "canonize" begin
+            using Qrochet: isleftcanonical, isrightcanonical
+
+            qtn = Chain(State(), Open(), [rand(4, 4), rand(4, 4, 4), rand(4, 4, 4), rand(4, 4, 4), rand(4, 4)])
+            canonized = canonize(qtn)
+
+            @test length(tensors(canonized)) == 9 # 5 tensors + 4 singular values vectors
+            @test isapprox(
+                contract(transform(TensorNetwork(canonized), Tenet.HyperindConverter())),
```
Comment (mofeing) [inline test/Ansatz/Chain_test.jl:158]: I think you no longer need to do the Tenet.HyperindConverter transformation to be able to contract. Would you mind updating your Tenet installation and trying it?

File: src/Ansatz/Chain.jl [hunk 2]
```diff
@@ -343,6 +343,51 @@ function isrightcanonical(qtn::Chain, site; atol::Real = 1e-12)
...
+canonize!(tn::Chain, args...; kwargs...) = canonize!(boundary(tn), tn, args...; kwargs...)
+
+"""
+canonize(boundary::Boundary, tn::Chain)
+
+Transform a `Chain` tensor network into the canonical form (Vidal form), that is,
+we have the singular values matrix Λᵢ between each tensor Γᵢ₋₁ and Γᵢ.
+"""
+function canonize!(::Open, tn::Chain)
+    S_vals = Tensor[]
```
Comment (mofeing) [inline src/Ansatz/Chain.jl:356]: Maybe we can just rename it to `Λ`?
```suggestion
    Λ = Tensor[]
```

File: src/Ansatz/Chain.jl [hunk 3]
```diff
@@ -343,6 +343,51 @@ function isrightcanonical(qtn::Chain, site; atol::Real = 1e-12)
...
+    for i in nsites(tn):-1:2
+        canonize_site!(tn, Site(i); direction = :left, method = :svd)
+
+        # extract the singular values and contract them with the next tensor
+        Λᵢ₋₁ = pop!(TensorNetwork(tn), select(tn, :between, Site(i - 1), Site(i)))
+        Bᵢ₋₁ = select(tn, :tensor, Site(i - 1))
+        replace!(TensorNetwork(tn), Bᵢ₋₁ => contract(Bᵢ₋₁, Λᵢ₋₁, dims = ()))
+        push!(S_vals, Λᵢ₋₁)
+    end
+    reverse!(S_vals) # reverse the singular values to match the order of the tensors
```
Comment (mofeing) [inline src/Ansatz/Chain.jl:373]: You can change the order:
1. right-to-left QR sweep
2. left-to-right SVD sweep

This way you don't need the `reverse!`

File: test/Ansatz/Chain_test.jl [hunk 4]
```diff
@@ -147,6 +147,45 @@
             @test length(tensors(canonize_site(qtn, Site(2); direction = :left, method = :svd))) == 4
         end
 
+        @testset "canonize" begin
+            using Qrochet: isleftcanonical, isrightcanonical
+
+            qtn = Chain(State(), Open(), [rand(4, 4), rand(4, 4, 4), rand(4, 4, 4), rand(4, 4, 4), rand(4, 4)])
```
Comment (mofeing) [inline test/Ansatz/Chain_test.jl:153]:
```suggestion
            qtn = MPS([rand(4, 4), rand(4, 4, 4), rand(4, 4, 4), rand(4, 4, 4), rand(4, 4)])
```

File: test/Ansatz/Chain_test.jl (hunk 1 above)
Comment (jofrevalles) [reply test/Ansatz/Chain_test.jl:158]: I now tried using `Tenet v0.5.2` and it did not work without the `Tenet.HyperindConverter`.

File: src/Ansatz/Chain.jl [hunk 5]
```diff
@@ -343,6 +343,51 @@ function isrightcanonical(qtn::Chain, site; atol::Real = 1e-12)
     return isapprox(contracted, identity_matrix; atol)
 end
 
+canonize(tn::Chain, args...; kwargs...) = canonize!(deepcopy(tn), args...; kwargs...)
```
Comment (mofeing) [inline src/Ansatz/Chain.jl:346]: Do you really need `deepcopy`? This copies all array contents too so it's costly. Doesn't it work with `copy`?

`deepcopy` should be just used in the case that the arrays of a TN are modified in-place. If the TN is modified in-place but the arrays not, then use `copy`.

**Comment by jofrevalles** [comment]
@mofeing The code now will error due to the `Tenet.HyperindConverter` thingy. I have tested locally and everything works otherwise.

**Pull request merged**